### ISSUE TITLE
ControlMessage: SOL_TLS control message falls back to ControlMessageOwned::Unknown

### ISSUE THREAD

**Issue by fasterthanlime** [nix-rust/nix#2064]
When enabling kTLS on a socket (which mostly involves [setting up the SOL_TLS sockopt](https://github.com/hapsoc/ktls/blob/c6d79c1f1f084ab52937723f1bbb88e4d503abee/src/ffi.rs#L184-L192)), that socket may receive some control messages like so:

```
  2023-06-29T11:56:33.154070Z TRACE ktls::ktls_stream: received UnknownCmsg(
    cmsghdr {
        cmsg_len: 17,
        cmsg_level: 282,
        cmsg_type: 2,
    },
    [
        23,
    ],
)
```

The cmsg_level corresponds to `libc::SOL_TLS`, cmsg_type corresponds to [TLS_GET_RECORD_TYPE](https://github.com/hapsoc/ktls-sys/blob/16149163e208dfbdd8789c40633348902bd6f991/src/bindings.rs#L80)

I believe the additional data in the `Vec<u8>` corresponds to TLS message types: 22 is handshake, 23 is application data, see [RFC 5246](https://datatracker.ietf.org/doc/html/rfc5246#section-6.2.1).

This happens because, after the TLS 1.3 handshake, tickets can be sent by the server to the client (for resumption, see [rustls 0.21's flag for that](https://docs.rs/rustls/0.21.2/rustls/server/struct.ServerConfig.html#structfield.send_tls13_tickets)).

If you're doing userland TLS, your deframer handles those messages gracefully and either throws away or stores the tickets for further use. But with kTLS, they have to go _somewhere_, and.. if you've received a ticket and you do a `read()` on the socket, you get [errno 5](https://github.com/hapsoc/ktls/issues/24) (helpful, I know!). If you do a `recvmsg()` you get one of these with the "bonus data" set to 22, which again, is the TLS 1.2/1.3 "handshake" type.

I'm working on a PR that recognizes those message types, adding a variant to `ControlMessageOwned`. The diff is probably a handful of lines and y'all don't need to bother yourselves with those details anyway but since there's fuck-all documentation about kTLS offline I needed to put it down _somewhere_.

**Comment by fasterthanlime** [comment]
> cmsg_type corresponds to `libc::SCM_CREDENTIALS`

I mean, they're both equal to 2, but that doesn't really say anything. I should probably dig deeper to find out what `cmsg_type` _really_ corresponds to when level is 282.

**Comment by Arnavion** [comment]
@fasterthanlime Are you going to upstream the `setsockopt` part of the kTLS setup to nix too? I've been working on my own impl at https://github.com/arsing/ktls-test and have a bunch of extensions to [libc](https://github.com/arsing/ktls-test/blob/main/libcext/src/lib.rs) and [nix](https://github.com/arsing/ktls-test/blob/main/nixext/src/setsockopt.rs) for it. Eg I noticed hapsoc/ktls is [doing it directly](https://github.com/hapsoc/ktls/blob/b61a868bbe88f5fd0d265f0bfa91b6262675b179/src/ffi.rs#L184-L192) whereas mine integrates into the nix `SetSockOpt` mechanism.

I was going to start PRing them, but since you've already done part of the effort maybe you want to do it.

**Comment by fasterthanlime** [comment]
@Arnavion feel free to go ahead with your own PRs, I’d be delighted to have it but am having enough trouble landing that one PR as is 🙃

**Comment by Arnavion** [comment]
Okay, thanks.

**Comment by asomers** [comment]
Fixed by #2065
